aten/src/ATen/native/cuda/BatchLinearAlgebra.cu

```diff
@@ -1893,6 +1951,145 @@ std::tuple<Tensor, Tensor> _triangular_solve_helper_cuda(const Tensor& self, con
+    MAGMAQueue magma_queue(self.get_device());
+
+    constexpr int64_t batch_limit = 65535;
+    // Compute as many batches of 65535 possible
+    // The number of "mini"-batches are floor(batchsize / batch_limit)
+    // and these cover floor(batchsize / batch_limit) * batch_limit
+    int64_t mini_batches = batchsize / batch_limit, mini_idx;
+    for (mini_idx = 0; mini_idx < mini_batches * batch_limit; mini_idx += batch_limit) {
+      scalar_t** self_array_cur = &self_array[mini_idx];
+      scalar_t** tau_array_cur = &tau_array[mini_idx];
+      scalar_t** T_array_cur = &T_array[mini_idx];
+      magmaLarftBatched<scalar_t>(m, k, 0, self_array_cur, lda, tau_array_cur, T_array_cur, ldt, &work_data, lwork, batch_limit, magma_queue);
+    }
+
+    // Compute whatever is left = batchsize - floor(batchsize / batch_limit) * batch_limit
+    // which concisely is equal to batchsize % batch_limit
+    if (batchsize % batch_limit != 0) {
+      magmaLarftBatched<scalar_t>(m, k, 0, &self_array[mini_idx], lda, &tau_array[mini_idx], &T_array[mini_idx], ldt, &work_data, lwork, batchsize % batch_limit, magma_queue);
+    }
+  }
+
+  // Compute result = (I - V T V^H) Ic,
+  // where I is an m x m identity matrix, Ic is an m x n identity matrix,
+  // V is an m x k matrix of reflectors and T is a k x k matrix
+  // With LAPACK same result could be obtained using LARFB
+  auto V = at::narrow(self, /*dim=*/-1, /*start=*/0, /*length=*/k);
+  auto W = at::narrow_copy(V, /*dim=*/-2, /*start=*/0, /*length=*/n);  // W <- Ic^H V: n x m @ m x k = n x k
+  at::matmul_out(W, T, W.conj().transpose(-2, -1));  // W <- T W^H: k x k @ k x n = k x n
+  at::matmul_out(self, V, W);  // self <- V W: m x k @ k x n = m x n
+  self.mul_(-1);
+  self.diagonal(/*offset=*/0, /*dim1=*/-2, /*dim2=*/-1).add_(1); // self <- self + Ic: m x n + m x n = m x n
+
+  // infos is not used here, make sure that it's filled with zeros
+  infos.fill_(0);
+#endif
+}
+
+// This is a type dispatching helper function for 'apply_orgqr'
+Tensor& orgqr_helper_batched_magma(Tensor& result, const Tensor& tau, Tensor& infos, int64_t n_columns) {
+  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES(result.scalar_type(), "orgqr_cuda", [&]{
+    apply_orgqr_batched_magma<scalar_t>(result, tau, infos, n_columns);
+  });
+  return result;
+}
+
+Tensor& orgqr_kernel_impl(Tensor& result, const Tensor& tau, Tensor& infos, int64_t n_columns) {
+#if defined(USE_CUSOLVER) && defined(USE_MAGMA)
+  if (batchCount(result) > 1 && tau.size(-1) <= 32) {
+    return orgqr_helper_batched_magma(result, tau, infos, n_columns); // magma-cuda
+  } else {
+    return orgqr_helper_cuda_lib(result, tau, infos, n_columns); // cusolver
+  }
+#elif defined(USE_CUSOLVER) && !defined(USE_MAGMA)
+  return orgqr_helper_cuda_lib(result, tau, infos, n_columns); // cusolver
+#else
+  TORCH_CHECK(tau.size(-1) <= 32, "orgqr: tau.shape[-1] larger than 32 is not supported with MAGMA backend. Please use PyTorch built with cuSOLVER support.")
```

Comment:
We should remember to change the function's signature and rename the second input as tau to clarify these references. (This proposal may be BC-breaking, but it's incredibly unlikely anyone is relying on the current names.)